editor/editor/layout.js

```diff
@@ -0,0 +1,67 @@
+/**
+ * Internal dependencies
+ */
+import ModeSwitcher from './mode-switcher';
+import EditorText from './mode/text';
+import EditorVisual from './mode/visual';
+
+class Layout extends wp.element.Component {
+	constructor() {
+		super( ...arguments );
+		this.switchMode = this.switchMode.bind( this );
+		this.changeHtml = this.changeHtml.bind( this );
+		this.changeBlocks = this.changeBlocks.bind( this );
+		this.state = {
+			mode: 'visual',
+			html: '',
+			blocks: []
+		};
+	}
+
+	componentDidMount() {
+		this.setState( {
+			html: this.props.content,
+			blocks: wp.blocks.parse( this.props.content )
+		} );
+	}
+
+	switchMode( newMode ) {
+		// TODO: we need a serializer from blocks here
+		const html = this.state.html;
+		const blocks = this.mode === 'visual' ? this.state.blocks : wp.blocks.parse( this.state.html );
+		this.setState( {
+			mode: newMode,
+			html,
+			blocks
+		} );
+	}
+
+	changeHtml( html ) {
+		this.setState( {
+			html
+		} );
+	}
+
+	changeBlocks( blocks ) {
+		this.setState( {
+			blocks
+		} );
+	}
+
+	render() {
+		const { mode, html, blocks } = this.state;
+		return (
+			<div>
+				<div className="editor__header">
+					<ModeSwitcher mode={ mode } onSwitch={ this.switchMode } />
+				</div>
+				<div className="editor__body">
+					{ mode === 'text' && <EditorText html={ html } onChange={ this.changeHtml } /> }
```

Comment:
Minor: Possible to consolidate similar logic, though dunno how much this helps readability:

```js
createChangeHandler( type ) {
	return ( value ) => this.setState( { [ type ]: value } );
}

// ...

render() {
	<EditorText html={ html } onChange={ this.createChangeHandler( 'html' ) } />
}

```